### ISSUE TITLE
Possible source of lockups with GangaObject._getWriteAccess

### ISSUE THREAD

**Issue by drmarkwslater** [ganga-devs/ganga#374]
I've just fixed a lockup in Tasks where the Task Monitor thread but I'm concerned this could cause problems in the future. The issue is this:

1. Object (Job or Task) is NOT loaded
2. thread calls `obj._getWriteAccess` which explicitly gets the `const_lock`
3. During this, it also starts to load the object which at various times will try to get the `internal_lock`
4. In another thread, the object is accessed which grabs the `internal_lock`
5. After this, it calls `obj._getReadAccess` attempt to load.
6. This then tries to grab the `const_lock`.

The order is not consistent in the two cases and so can lead to deadlocks.

This was a problem in Tasks because of the specific call to `_getWriteAccess` but I'm concerned this could happen elsewhere. I'm not sure how to proceed with this but I thought I'd put in an issue so we remember :smile:

**Comment by rob-c** [comment]
I've been fighting with the Access locks elsewhere and the Registry simply stops doing them during Registry transactions which isn't ideal.

I think a _simple_ approach would be to abandon the read/write lock for explicitly calling a load of the job and disallowing multiple ganga sessions from accessing an object loaded by another ganga session? Do we really want/need the complexity that we have with the read/write access locks? (I really don't know here so if we do then this is fine)

**Comment by milliams** [comment]
The read/write access locks (maintained as physical files on disk) are useful/necessary if we want to support two active Ganga session accessing the same repository. I'm not convinced that this is a feature that we want to support to this extent and without a good set of usecases/user-stories it's more of a burden than anything. I would be in favour of removing per-object session locks until we can reimplement it properly once we know what it's actually useful for. Someone needs to write down a set of situations/use-cases where it would be useful before we can really ascertain what level of complexity is warranted.

However, having a single global session lock is probably still useful to inform the user if there's still another Ganga session running.

**Comment by rob-c** [comment]
Given that the locks at present prevent 2 sessions from modifying the same job I would rather not remove it without completely removing the ability to have 2 Ganga sessions.

Warning the user isn't enough as some(most?) don't pay attention to the warnings on startup until they run into more serious issues later on such as corrupting their job with 1k subjobs.

We still have to put in a mechanism to disable the monitoring if we detect another Ganga session is already running btw which was brought up in another PR a while ago.

**Comment by egede** [comment]
The use case for having two simultaneous Ganga sessions is actually quite common. There is a script (or an interactive command) that submits a very large job in one Ganga session and then a second Ganga session that is monitoring jobs.

**Comment by rob-c** [comment]
For now simply deferring the _getReadAccess and _getWriteAccess calls on a job which is currently being loaded from the Registry in the load method should alleviate this problem completely and should be what the repo is doing.

**Comment by rob-c** [comment]
Given we no longer have this method can we close this as no longer valid? I suspect we've likely fixed whatever this was in the meantime.

**Comment by drmarkwslater** [comment]
Not certain but it should be fixed. I'll try to have a closer look in a few days...

**Comment by drmarkwslater** [comment]
Fixed by #627. The `getWriteAccess` and `getReadAccess` methods have been replaced and the explicit acquiring of the object `const_lock` no longer happens.
